pinot-query-runtime/src/main/java/org/apache/pinot/query/service/server/QueryServer.java

```diff
@@ -208,43 +225,61 @@ public void explain(Worker.QueryRequest request, StreamObserver<Worker.ExplainRe
+                  List<Worker.WorkerMetadata> protoWorkerMetadataList =
+                      QueryPlanSerDeUtils.toProtoWorkerMetadataList(metadata.getWorkerMetadataList());
 
-                StageMetadata metadata = plan.getStageMetadata();
-                List<Worker.WorkerMetadata> protoWorkerMetadataList =
-                    QueryPlanSerDeUtils.toProtoWorkerMetadataList(metadata.getWorkerMetadataList());
+                  builder.addStagePlan(Worker.StagePlan.newBuilder().setRootNode(rootAsBytes).setStageMetadata(
+                      Worker.StageMetadata.newBuilder().setStageId(metadata.getStageId())
+                          .addAllWorkerMetadata(protoWorkerMetadataList)
+                          .setCustomProperty(QueryPlanSerDeUtils.toProtoProperties(metadata.getCustomProperties()))));
+                }
+                builder.putMetadata(CommonConstants.Explain.Response.ServerResponseStatus.STATUS_OK, "");
+                synchronized (responseObserver) {
+                  responseObserver.onNext(builder.build());
+                }
+                return null;
+              })
+          ).toArray(CompletableFuture[]::new);
 
-                builder.addStagePlan(Worker.StagePlan.newBuilder().setRootNode(rootAsBytes).setStageMetadata(
-                    Worker.StageMetadata.newBuilder().setStageId(metadata.getStageId())
-                        .addAllWorkerMetadata(protoWorkerMetadataList)
-                        .setCustomProperty(QueryPlanSerDeUtils.toProtoProperties(metadata.getCustomProperties()))));
-              }
-              builder.putMetadata(CommonConstants.Explain.Response.ServerResponseStatus.STATUS_OK, "");
-              responseObserver.onNext(builder.build());
-            });
-      } catch (ExecutionException | InterruptedException | TimeoutException | RuntimeException e) {
-        long requestId = QueryThreadContext.getRequestId();
-        LOGGER.error("Caught exception while submitting request: {}", requestId, e);
-        String errorMsg = "Caught exception while submitting request: " + e.getMessage();
-        responseObserver.onNext(Worker.ExplainResponse.newBuilder()
-            .putMetadata(CommonConstants.Explain.Response.ServerResponseStatus.STATUS_ERROR, errorMsg)
+      // A completable future that will finish when all submit task finish or on timoeut
+      CompletableFuture<Void> allCompleted = CompletableFuture.allOf(responseFutures)
+          .orTimeout(QueryThreadContext.getDeadlineMs(), TimeUnit.MILLISECONDS);
+      // When this future completes, notify the broker.
+      allCompleted.handle((result, error) -> {
+        if (error != null) {
+          long requestId = QueryThreadContext.getRequestId();
+          LOGGER.error("Caught exception while submitting request: {}", requestId, error);
+          String errorMsg = "Caught exception while submitting request: " + error.getMessage();
+          synchronized (responseObserver) {
```

Comment:
Nice catch, I wasn't aware that `StreamObserver` doesn't guarantee thread safety.